# Beyond Hello World, A Computer Vision Example
In the previous exercise you saw how to create a neural network that figured out the problem you were trying to solve. This gave an explicit example of learned behavior. Of course, in that instance, it was a bit of overkill because it would have been easier to write the function Y=2x-1 directly, instead of bothering with using Machine Learning to learn the relationship between X and Y for a fixed set of values, and extending that for all values.

But what about a scenario where writing rules like that is much more difficult -- for example a computer vision problem? Let's take a look at a scenario where we can recognize different items of clothing, trained from a dataset containing 10 different types.

## Start Coding

Let's start with our import of TensorFlow

In [1]:
import tensorflow as tf
print(tf.__version__)

C:\Users\surface\ananew\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


1.13.1


The Fashion MNIST data is available directly in the tf.keras datasets API. You load it like this:

In [2]:
mnist = tf.keras.datasets.fashion_mnist

Calling load_data on this object will give you two sets of two lists, these will be the training and testing values for the graphics that contain the clothing items and their labels.


In [3]:
(training_images, training_labels), (test_images, test_labels) = mnist.load_data()

4423680/4422102 [==============================] - 0s 0us/step


What does these values look like? Let's print a training image, and a training label to see...Experiment with different indices in the array. For example, also take a look at index 42...that's a a different boot than the one at index 0


In [4]:
training_images.shape[0]

60000

In [5]:
training_images.shape[1] # why is not 28*28??

28

In [6]:
import matplotlib.pyplot as plt
plt.imshow(training_images[0])
print(training_labels[0])
print(training_images[0])

9
[[  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   1   0   0  13  73   0
    0   1   4   0   0   0   0   1   1   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   3   0  36 136 127  62
   54   0   0   0   1   3   4   0   0   3]
 [  0   0   0   0   0   0   0   0   0   0   0   0   6   0 102 204 176 134
  144 123  23   0   0   0   0  12  10   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0   0 155 236 207 178
  107 156 161 109  64  23  77 130  72  15]
 [  0   0   0   0   0   0   0   0   0   0   0   1   0  69 207 223 218 216
  216 163 127 121 122 146 141  88 172  66]
 [  0   0   0   0   0   0   0   0   0   1   1   1   0 200 232 

You'll notice that all of the values in the number are between 0 and 255. If we are training a neural network, for various reasons it's easier if we treat all values as between 0 and 1, a process called '**normalizing**'...and fortunately in Python it's easy to normalize a list like this without looping. You do it like this:

In [7]:
training_images  = training_images / 255.0
test_images = test_images / 255.0

Now you might be wondering why there are 2 sets...training and testing -- remember we spoke about this in the intro? The idea is to have 1 set of data for training, and then another set of data...that the model hasn't yet seen...to see how good it would be at classifying values. After all, when you're done, you're going to want to try it out with data that it hadn't previously seen!

Let's now design the model. There's quite a few new concepts here, but don't worry, you'll get the hang of them. 

In [9]:
model = tf.keras.models.Sequential([tf.keras.layers.Flatten(), 
                                    tf.keras.layers.Dense(128, activation=tf.nn.relu), 
                                    tf.keras.layers.Dense(10, activation=tf.nn.softmax)])

**Sequential**: That defines a SEQUENCE of layers in the neural network

**Flatten**: Remember earlier where our images were a square, when you printed them out? Flatten just takes that square and turns it into a 1 dimensional set.

**Dense**: Adds a layer of neurons

Each layer of neurons need an **activation function** to tell them what to do. There's lots of options, but just use these for now. 

**Relu** effectively means "If X>0 return X, else return 0" -- so what it does it it only passes values 0 or greater to the next layer in the network.

**Softmax** takes a set of values, and effectively picks the biggest one, so, for example, if the output of the last layer looks like [0.1, 0.1, 0.05, 0.1, 9.5, 0.1, 0.05, 0.05, 0.05], it saves you from fishing through it looking for the biggest value, and turns it into [0,0,0,0,1,0,0,0,0] -- The goal is to save a lot of coding!


The next thing to do, now the model is defined, is to actually build it. You do this by compiling it with an optimizer and loss function as before -- and then you train it by calling **model.fit ** asking it to fit your training data to your training labels -- i.e. have it figure out the relationship between the training data and its actual labels, so in future if you have data that looks like the training data, then it can make a prediction for what that data would look like. 

In [10]:
model.compile(optimizer = tf.train.AdamOptimizer(),
              loss = 'sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(training_images, training_labels, epochs=100)

Instructions for updating:
Colocations handled automatically by placer.
Epoch 1/100
60000/60000 [==============================] - 6s 97us/sample - loss: 0.4987 - acc: 0.8243
Epoch 2/100
60000/60000 [==============================] - 5s 82us/sample - loss: 0.3728 - acc: 0.8649
Epoch 3/100
60000/60000 [==============================] - 5s 83us/sample - loss: 0.3358 - acc: 0.8770
Epoch 4/100
60000/60000 [==============================] - 6s 97us/sample - loss: 0.3144 - acc: 0.8849
Epoch 5/100
60000/60000 [==============================] - 5s 91us/sample - loss: 0.2950 - acc: 0.8895
Epoch 6/100
60000/60000 [==============================] - 5s 87us/sample - loss: 0.2795 - acc: 0.89580s - loss: 0.2799 - acc: 0
Epoch 7/100
60000/60000 [==============================] - 7s 122us/sample - loss: 0.2688 - acc: 0.8993
Epoch 8/100
60000/60000 [==============================] - 7s 110us/sample - loss: 0.2574 - acc: 0.9033
Epoch 9/100
60000/60000 [==============================] - 6s 97us/sample - 

Once it's done training -- you should see an accuracy value at the end of the final epoch. It might look something like 0.9098. This tells you that your neural network is about 91% accurate in classifying the training data. I.E., it figured out a pattern match between the image and the labels that worked 91% of the time. Not great, but not bad considering it was only trained for 5 epochs and done quite quickly.

But how would it work with unseen data? That's why we have the test images. We can call model.evaluate, and pass in the two sets, and it will report back the loss for each. Let's give it a try:

In [11]:
model.evaluate(test_images, test_labels)

10000/10000 [==============================] - 1s 50us/sample - loss: 0.6732 - acc: 0.8889


[0.6732343255124986, 0.8889]

For me, that returned a accuracy of about .8889, which means it was about 88% accurate. As expected it probably would not do as well with *unseen* data as it did with data it was trained on!  As you go through this course, you'll look at ways to improve this. 

To explore further, try the below exercises:


# Exploration Exercises

### Exercise 1:

For this first exercise run the below code: It creates a set of classifications for each of the test images, and then prints the first entry in the classifications. The output, after you run it is a list of numbers. Why do you think this is, and what do those numbers represent? 

In [12]:
classifications = model.predict(test_images)

print(classifications[0]) # the prediction of the first example in testing dataset

[6.8797893e-24 0.0000000e+00 1.1207443e-33 0.0000000e+00 8.5451968e-32
 6.2038070e-18 3.3223834e-32 1.0266690e-07 6.8054532e-23 9.9999988e-01]


<font color = 'green'>
Maybe the nums above represent the mean predicted classified numbers of the pictures in each classification
<font color = 'red'>
wrong

Hint: try running print(test_labels[0]) -- and you'll get a 9. Does that help you understand why this list looks the way it does? 

In [13]:
print(test_labels[0])

9


<font color = 'green'>
There're 10 labels in this example in total

### What does this list represent?

1.   It's 10 random meaningless values
2.   It's the first 10 classifications that the computer made
<font color = 'red'>
3.   It's the probability that this item is each of the 10 classes

#### Answer: 

The correct answer is (3)

The output of the model is a list of 10 numbers. These numbers are a probability that the value being classified is the corresponding value, i.e. the first value in the list is the probability that the handwriting is of a '0', the next is a '1' etc. Notice that they are all VERY LOW probabilities.

For the 7, the probability was .999+, i.e. the neural network is telling us that it's almost certainly a 7.

### How do you know that this list tells you that the item is an ankle boot?


1.   There's not enough information to answer that question
<font color = 'red'>
2.   The probability of the 10th element on the list is the biggest, and the ankle boot is labelled 9
<font color = 'black'>
2.   The ankle boot is label 9, and there are 0->9 elements in the list




#### Answer

The correct answer is (2). 

Both the list and the labels are 0 based, so the ankle boot having label 9 means that it is the 10th of the 10 classes. The list having the 10th element being the highest value means that the Neural Network has predicted that the item it is classifying is most likely an ankle boot

## Exercise 2: 

Let's now look at the layers in your model. Experiment with different values for the dense layer with 512 neurons. 

What different results do you get for loss, training time etc? Why do you think that's the case? 

In [14]:
import tensorflow as tf
print(tf.__version__)

mnist = tf.keras.datasets.mnist

(training_images, training_labels) ,  (test_images, test_labels) = mnist.load_data()

training_images = training_images/255.0
test_images = test_images/255.0

model = tf.keras.models.Sequential([tf.keras.layers.Flatten(),
                                    tf.keras.layers.Dense(512, activation=tf.nn.relu),
                                    tf.keras.layers.Dense(10, activation=tf.nn.softmax)])

model.compile(optimizer = 'adam',
              loss = 'sparse_categorical_crossentropy')

model.fit(training_images, training_labels, epochs=5)

model.evaluate(test_images, test_labels)

classifications = model.predict(test_images)

print(classifications[0])
print(test_labels[0])

1.13.1
11493376/11490434 [==============================] - 1s 0us/step
Epoch 1/5
60000/60000 [==============================] - 23s 376us/sample - loss: 0.2026
Epoch 2/5
60000/60000 [==============================] - 23s 383us/sample - loss: 0.0808
Epoch 3/5
60000/60000 [==============================] - 25s 421us/sample - loss: 0.0527
Epoch 4/5
60000/60000 [==============================] - 24s 392us/sample - loss: 0.0360
Epoch 5/5
10000/10000 [==============================] - 1s 112us/sample - loss: 0.0598
[8.5191736e-11 1.0198766e-09 2.4212254e-08 3.7656766e-05 1.0017182e-12
 2.1229322e-08 5.9425589e-14 9.9996233e-01 5.6192588e-09 2.4714769e-08]
7


<font color = 'green'>
We can see from the processing time(which is significantly longer) and the accuracy of testing dataset(which improves a lot -- 0.0598 loss compared to 0.6732 loss) that deep neural network with more units works bettwe than shallow neural network with fewer units. That is to say, the neural network with more hidden layers and hidden units is better than the ones with fewer hidden layers and hidden units.

Also, the classification prediction of the first example in testing dataset is different from the previous neural network model 

### Question 1. Increase to 1024 Neurons -- What's the impact?
<font color = 'red'>
1. Training takes longer, but is more accurate
<font color = 'black'>
2. Training takes longer, but no impact on accuracy
3. Training takes the same time, but is more accurate


#### Answer

The correct answer is (1) by adding more Neurons we have to do more calculations, slowing down the process, but in this case they have a good impact -- we do get more accurate. That doesn't mean it's always a case of 'more is better', you can hit the law of diminishing returns very quickly!

## Exercise 3: 

What would happen if you remove the Flatten() layer. Why do you think that's the case? 

You get an error about the shape of the data. It may seem vague right now, but it **reinforces the rule of thumb** that **the first layer in your network should be the same shape as your data**. Right now our data is 28x28 images, and 28 layers of 28 neurons would be infeasible, so it makes more sense to 'flatten' that 28,28 into a 784x1. 
<font color = 'green'>
Instead of wriitng all the code to handle that ourselves, we add the Flatten() layer at the begining, and when the arrays are loaded into the model later, they'll automatically be flattened for us.

In [15]:
import tensorflow as tf
print(tf.__version__)

mnist = tf.keras.datasets.mnist

(training_images, training_labels) ,  (test_images, test_labels) = mnist.load_data()

training_images = training_images/255.0
test_images = test_images/255.0

model = tf.keras.models.Sequential([#tf.keras.layers.Flatten(),
                                    tf.keras.layers.Dense(64, activation=tf.nn.relu),
                                    tf.keras.layers.Dense(10, activation=tf.nn.softmax)])

model.compile(optimizer = 'adam',
              loss = 'sparse_categorical_crossentropy')

model.fit(training_images, training_labels, epochs=5)

model.evaluate(test_images, test_labels)

classifications = model.predict(test_images)

print(classifications[0])
print(test_labels[0])

1.13.1
Epoch 1/5


InvalidArgumentError: logits and labels must have the same first dimension, got logits shape [896,10] and labels shape [32]
	 [[{{node loss_2/output_1_loss/SparseSoftmaxCrossEntropyWithLogits/SparseSoftmaxCrossEntropyWithLogits}}]]

## Exercise 4: 

Consider the final (output) layers. Why are there 10 of them? What would happen if you had a different amount than 10? For example, try training the network with 5

You get an error as soon as it finds an unexpected value. **Another rule of thumb**-- **the number of neurons in the last layer should match the number of classes you are classifying for**. In this case it's the digits 0-9, so there are 10 of them, hence you should have 10 neurons in your final layer.

In [17]:
import tensorflow as tf
print(tf.__version__)

mnist = tf.keras.datasets.mnist

(training_images, training_labels) ,  (test_images, test_labels) = mnist.load_data()

training_images = training_images/255.0
test_images = test_images/255.0

model = tf.keras.models.Sequential([tf.keras.layers.Flatten(),
                                    tf.keras.layers.Dense(64, activation=tf.nn.relu),
                                    tf.keras.layers.Dense(5, activation=tf.nn.softmax)])

model.compile(optimizer = 'adam',
              loss = 'sparse_categorical_crossentropy')

model.fit(training_images, training_labels, epochs=5)

model.evaluate(test_images, test_labels)

classifications = model.predict(test_images)

print(classifications[0])
print(test_labels[0])

1.13.1
Epoch 1/5


InvalidArgumentError: Received a label value of 9 which is outside the valid range of [0, 5).  Label values: 2 8 1 7 1 4 2 9 0 5 5 3 7 5 8 4 5 1 4 5 6 8 2 9 8 9 5 9 0 5 1 2
	 [[{{node loss_4/output_1_loss/SparseSoftmaxCrossEntropyWithLogits/SparseSoftmaxCrossEntropyWithLogits}}]]

## Exercise 5: 

Consider the effects of the amount of units in hidden layers in the network. What will happen if you add another layer between the one with 512 and the final layer with 10. 

Ans: There isn't a significant impact -- because this is relatively simple data. For far more complex data (including color images to be classified as flowers that you'll see in the next lesson), extra layers are often necessary. 

In [19]:
import tensorflow as tf
print(tf.__version__)

mnist = tf.keras.datasets.mnist

(training_images, training_labels) ,  (test_images, test_labels) = mnist.load_data()

training_images = training_images/255.0
test_images = test_images/255.0

model = tf.keras.models.Sequential([tf.keras.layers.Flatten(),
                                    tf.keras.layers.Dense(512, activation=tf.nn.relu),
                                    tf.keras.layers.Dense(256, activation=tf.nn.relu),
                                    tf.keras.layers.Dense(10, activation=tf.nn.softmax)])

model.compile(optimizer = 'adam',
              loss = 'sparse_categorical_crossentropy')

model.fit(training_images, training_labels, epochs=5)

model.evaluate(test_images, test_labels)

classifications = model.predict(test_images)

print(classifications[0])
print(test_labels[0])

1.13.1
Epoch 1/5
60000/60000 [==============================] - 29s 476us/sample - loss: 0.1868
Epoch 2/5
60000/60000 [==============================] - 28s 469us/sample - loss: 0.0801
Epoch 3/5
60000/60000 [==============================] - 29s 483us/sample - loss: 0.0535
Epoch 4/5
60000/60000 [==============================] - 27s 454us/sample - loss: 0.0419
Epoch 5/5
10000/10000 [==============================] - 1s 141us/sample - loss: 0.0727
[4.5655229e-12 1.1329102e-08 1.3262921e-10 1.9695817e-07 2.3211833e-11
 3.3863305e-12 1.6943873e-14 9.9999976e-01 7.5870539e-11 6.9224488e-09]
7


## Exercise 6: 

Consider the impact of training for more or less epochs. Why do you think that would be the case? 

Try 15 epochs -- you'll probably get a model with a worse loss of testing dataset than the one with 5 epochs -- **you might see the loss value stops decreasing, and sometimes increases. This is a side effect of something called 'overfitting' which you can learn about [somewhere] and it's something you need to keep an eye out for when training neural networks**. There's no point in wasting your time training if you aren't improving your loss, right! :)

In [21]:
import tensorflow as tf
print(tf.__version__)

mnist = tf.keras.datasets.mnist

(training_images, training_labels) ,  (test_images, test_labels) = mnist.load_data()

training_images = training_images/255.0
test_images = test_images/255.0

model = tf.keras.models.Sequential([tf.keras.layers.Flatten(),
                                    tf.keras.layers.Dense(512, activation=tf.nn.relu),
                                    tf.keras.layers.Dense(10, activation=tf.nn.softmax)])

model.compile(optimizer = 'adam',
              loss = 'sparse_categorical_crossentropy')

model.fit(training_images, training_labels, epochs=15)

model.evaluate(test_images, test_labels)

classifications = model.predict(test_images)

print(classifications[34])
print(test_labels[34])

1.13.1
Epoch 1/15
60000/60000 [==============================] - 24s 396us/sample - loss: 0.1990
Epoch 2/15
60000/60000 [==============================] - 24s 404us/sample - loss: 0.0804
Epoch 3/15
60000/60000 [==============================] - 25s 414us/sample - loss: 0.0523
Epoch 4/15
60000/60000 [==============================] - 26s 431us/sample - loss: 0.0374
Epoch 5/15
60000/60000 [==============================] - 26s 432us/sample - loss: 0.0272
Epoch 6/15
60000/60000 [==============================] - 25s 420us/sample - loss: 0.0212
Epoch 7/15
60000/60000 [==============================] - 25s 416us/sample - loss: 0.0175
Epoch 8/15
60000/60000 [==============================] - 25s 412us/sample - loss: 0.0139
Epoch 9/15
60000/60000 [==============================] - 24s 395us/sample - loss: 0.0132
Epoch 10/15
60000/60000 [==============================] - 26s 426us/sample - loss: 0.0118
Epoch 11/15
60000/60000 [==============================] - 25s 417us/sample - loss: 0.0098
E

<font color = 'green'>
The loss of the testing data when epochs equals to 5 is about 0.0598

## Exercise 7: 

Before you trained, you normalized the data, going from values that were 0-255 to values that were 0-1. What would be the impact of removing that? Here's the complete code to give it a try. Why do you think you get different results? 

<font color = 'green'>
train the model with unnormalized data

In [22]:
import tensorflow as tf
print(tf.__version__)
mnist = tf.keras.datasets.mnist
(training_images, training_labels), (test_images, test_labels) = mnist.load_data()

model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(128, activation=tf.nn.relu),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy')
model.fit(training_images, training_labels, epochs=5)
model.evaluate(test_images, test_labels)
classifications = model.predict(test_images)
print(classifications[0])
print(test_labels[0])

1.13.1
Epoch 1/5
60000/60000 [==============================] - 7s 120us/sample - loss: 9.2230
Epoch 2/5
60000/60000 [==============================] - 7s 121us/sample - loss: 7.8654
Epoch 3/5
60000/60000 [==============================] - 9s 155us/sample - loss: 7.6720
Epoch 4/5
60000/60000 [==============================] - 9s 152us/sample - loss: 7.5210
Epoch 5/5
10000/10000 [==============================] - 1s 67us/sample - loss: 7.2054
[0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]
7


<font color = 'green'>
train the model with unnormalized data

In [23]:
import tensorflow as tf
print(tf.__version__)
mnist = tf.keras.datasets.mnist
(training_images, training_labels), (test_images, test_labels) = mnist.load_data()
training_images=training_images/255.0
test_images=test_images/255.0
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(128, activation=tf.nn.relu),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy')
model.fit(training_images, training_labels, epochs=5)
model.evaluate(test_images, test_labels)
classifications = model.predict(test_images)
print(classifications[0])
print(test_labels[0])

1.13.1
Epoch 1/5
60000/60000 [==============================] - 8s 127us/sample - loss: 0.2596s - loss: 0.2 - ETA: 0s - los
Epoch 2/5
60000/60000 [==============================] - 7s 115us/sample - loss: 0.1128
Epoch 3/5
60000/60000 [==============================] - 6s 107us/sample - loss: 0.0774
Epoch 4/5
60000/60000 [==============================] - 7s 118us/sample - loss: 0.0580
Epoch 5/5
10000/10000 [==============================] - 1s 69us/sample - loss: 0.0682
[2.3136521e-07 4.7205134e-08 6.7421046e-05 1.5692001e-04 1.7808109e-10
 2.7456924e-06 1.8238536e-12 9.9968517e-01 2.4499877e-05 6.2899016e-05]
7


## Exercise 8: 

Earlier when you trained for extra epochs you had an issue where your loss might change. It might have taken a bit of time for you to wait for the training to do that, and you might have thought 'wouldn't it be nice if I could stop the training when I reach a desired value?' -- i.e. 95% accuracy might be enough for you, and if you reach that after 3 epochs, why sit around waiting for it to finish a lot more epochs....So how would you fix that? Like any other program...you have callbacks! Let's see them in action...

In [1]:
import tensorflow as tf
print(tf.__version__)

class myCallback(tf.keras.callbacks.Callback):
  def on_epoch_end(self, epoch, logs={}):
    if(logs.get('loss')<0.4):
      print("\nReached 60% accuracy so cancelling training!")
      self.model.stop_training = True

callbacks = myCallback()

mnist = tf.keras.datasets.fashion_mnist

(training_images, training_labels), (test_images, test_labels) = mnist.load_data()

training_images=training_images/255.0
test_images=test_images/255.0

model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(128, activation=tf.nn.relu),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy')
model.fit(training_images, training_labels, epochs=5, callbacks=[callbacks])


C:\Users\surface\ananew\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


1.13.1
Instructions for updating:
Colocations handled automatically by placer.
Epoch 1/5
60000/60000 [==============================] - 7s 125us/sample - loss: 0.4966
Epoch 2/5
59808/60000 [============================>.] - ETA: 0s - loss: 0.3744- ETA: 0s - l
Reached 60% accuracy so cancelling training!
60000/60000 [==============================] - 8s 129us/sample - loss: 0.3747
